### PR TITLE
Add argument mutating New type

### LINKED ISSUE #145: Create model mutator
A type to pass optional argument mutations.

### PR DESCRIPTION
**Pull request by amulware**
Closes #145.

We had some discussion as to the best name for this. Options like 'Mutator' were mentioned. I went with 'New' for brevity, and because it reads nicely.

The usage of the type would be as follows (we will add this to the models in another PR).

```
public Project With(
    New<int> projectId = default(New<int>),
    New<string> name = default(New<string>)
    )
    => new Project(
        projectId.ValueOr(ProjectId),
        name.ValueOr(Name)
    );
```

```
var renamedProject = project.With(name: "this is dog");
```

Also, while the type is possibly only going to be used in Foundation, I figured it was general and context-less enough to add to Multivac.

### CODE REVIEW

File: Toggl.Multivac/New.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        private readonly T value;
+
+        private New(T value)
+        {
+            hasValue = true;
+            this.value = value;
+        }
+
+        public T ValueOr(T oldValue) => hasValue ? value : oldValue;
+
```
Comment (willsb) [inline Toggl.Multivac/New.cs]: delet this

Comment (amulware) [inline Toggl.Multivac/New.cs]: I wanted to keep a clear destination before the static constructors/property. It does not follow guidelines in the first place with the property down here, but I think that None and Value should not be separated, and then I rather have them down here. You still don't want the extra empty line?

**Comment by amulware** [comment]
I did a quick google earlier, and I'm not aware of generic theories. In this case, we want to run a bunch of tests on the same several pieces of data as well, which might cause a lot of duplication.

I rather keep it like this.

**Comment by amulware** [comment]
Feel free to squash for me.

**Pull request merged**